test/src/test/java/org/corfudb/infrastructure/AutoCommitServiceTest.java

```diff
@@ -0,0 +1,146 @@
+                .addSequencer(SERVERS.PORT_0)
+                .addSequencer(SERVERS.PORT_1)
+                .addSequencer(SERVERS.PORT_2)
+                .buildSegment()
+                .setReplicationMode(Layout.ReplicationMode.CHAIN_REPLICATION)
+                .buildStripe()
+                .addLogUnit(SERVERS.PORT_0)
+                .addLogUnit(SERVERS.PORT_1)
+                .addLogUnit(SERVERS.PORT_2)
+                .addToSegment()
+                .addToLayout()
+                .build();
+
+        bootstrapAllServers(l);
+
+        // Shutdown management servers.
+        getManagementServer(SERVERS.PORT_0).shutdown();
+        getManagementServer(SERVERS.PORT_1).shutdown();
+        getManagementServer(SERVERS.PORT_2).shutdown();
+
+        return l;
+    }
+
+    private LogData getLogData(TokenResponse token, byte[] payload) {
+        LogData ld = new LogData(DataType.DATA, payload);
+        ld.useToken(token);
+        return ld;
+    }
+
+    private LogData read(CorfuRuntime runtime, long address, String server) throws Exception {
+        return runtime.getLayoutView().getRuntimeLayout().getLogUnitClient(server)
+                .read(address).get().getAddresses().get(address);
+    }
+
+    @Test
+    @SuppressWarnings("checkstyle:magicnumber")
+    public void testAutoCommitHoles() throws Exception {
+        CorfuRuntime runtime = getRuntime(setup3NodeCluster()).connect();
+
+        final int numIter = 100;
+        Set<Long> noWriteHoles = new HashSet<>(Arrays.asList(15L, 25L, 45L, 85L));
+        Set<Long> partialWriteHoles = new HashSet<>(Arrays.asList(55L, 75L));
+
+        for (long i = 0; i < numIter; i++) {
+            TokenResponse token = runtime.getSequencerView().next();
+            if (noWriteHoles.contains(i)) {
+                // Write nothing to create a hole on all log units.
+            } else if (partialWriteHoles.contains(i)) {
+                // Write to head log unit to create a partial write hole.
+                runtime.getLayoutView().getRuntimeLayout().getLogUnitClient(SERVERS.ENDPOINT_0)
+                        .write(getLogData(token, "partial write".getBytes())).get();
+            } else {
+                runtime.getAddressSpaceView().write(token, "Test Payload".getBytes());
+            }
+        }
+
+        AutoCommitService autoCommitService =
+                getManagementServer(SERVERS.PORT_0).getManagementAgent().getAutoCommitService();
+        // First invocation would only set the next commit end to the current log tail.
+        autoCommitService.runAutoCommit();
+
+        for (long i = 0; i < numIter; i++) {
+            if (noWriteHoles.contains(i) || partialWriteHoles.contains(i)) {
+                assertThat(read(runtime, i, SERVERS.ENDPOINT_2).getType()).isEqualTo(DataType.EMPTY);
+            } else {
+                assertThat(read(runtime, i, SERVERS.ENDPOINT_2).getType()).isEqualTo(DataType.DATA);
+            }
+        }
```

Comment:
Should you verify on getting the commit tail from all log unit servers?

Reply:
This is checked at the end.